### PR TITLE
Spanner DML/Partitioned DML

### PR DESCRIPTION
**Pull request by blowmage**
Adds DML and Partitioned DML support to the Spanner client and batch client.

```ruby
require "google/cloud/spanner"

spanner = Google::Cloud::Spanner.new
db = spanner.client "my-instance", "my-database"

db.transaction do |tx|
  row_count = tx.execute_update(
    "UPDATE users SET name = @name WHERE id = @id",
    params: { id: 1, name: "Charlie" }
  ) 
end
```

```ruby
require "google/cloud/spanner"

spanner = Google::Cloud::Spanner.new
db = spanner.client "my-instance", "my-database"

row_count = db.execute_partition_update(
  "UPDATE users SET friends = NULL WHERE active = @active",
  params: { active: false }
)
```

### CODE REVIEW

**Comment by blowmage** [comment]
FYI @snehashah16 @crwilcox

File: google-cloud-spanner/README.md [hunk 1]
```diff
@@ -28,7 +28,7 @@ spanner = Google::Cloud::Spanner.new
 db = spanner.client "my-instance", "my-database"
 
 db.transaction do |tx|
-  results = tx.execute "SELECT * FROM users"
+  results = tx.execute_query "SELECT * FROM users"
```
Comment (snehashah16) [inline google-cloud-spanner/README.md]: @blowmage - is there a possibility to leave this rename change out as per initial PR comment I had.
This is unrelated to DML / PDML and I rather not have a hiccup w/ the launch

**Comment by blowmage** [comment]
The acceptance tests pass while using a whitelisted project, but do not yet pass using the standard project used for CI, so this PR should not be merged until DML/PDML are generally available.

File: google-cloud-spanner/acceptance/spanner/client/large_data_test.rb [hunk 2]
```diff
@@ -85,7 +85,7 @@ def random_row
   it "writes and queries bytes" do
     my_row = random_row
     db.upsert table_name, my_row
-    results = db.execute "SELECT id, string, byte, strings, bytes FROM #{table_name} WHERE id = @id", params: { id: my_row[:id] }
+    results = db.execute_sql "SELECT id, string, byte, strings, bytes FROM #{table_name} WHERE id = @id", params: { id: my_row[:id] }
```
Comment (snehashah16) [inline google-cloud-spanner/acceptance/spanner/client/large_data_test.rb]: also its confusing whats the difference between execute_sql and execute_query?

Comment (blowmage) [reply google-cloud-spanner/acceptance/spanner/client/large_data_test.rb]: There is no difference, it is the same method with an aliased name. I changed some calls to `execute_sql` so that the alias was used in acceptance tests.

File: google-cloud-spanner/README.md (hunk 1 above)
Comment (blowmage) [reply google-cloud-spanner/README.md]: Yes. I can move it to a separate PR.

**Comment by blowmage** [comment]
This is ready to be reviewed.

File: google-cloud-spanner/lib/google/cloud/spanner/transaction.rb [hunk 3]
```diff
@@ -245,13 +246,109 @@ def transaction_id
...
+        #       params: { id: 1, name: "Charlie" }
+        #     )
+        #   end
+        #
+        def execute_update sql, params: nil, types: nil
+          results = execute sql, params: params, types: types
+          # Stream all PartialResultSet to get ResultSetStats
+          results.rows.to_a
+          # Raise an error if there is not a row count returned
+          raise Google::Cloud::InvalidArgumentError if results.row_count.nil?
```
Comment (quartzmo) [inline google-cloud-spanner/lib/google/cloud/spanner/transaction.rb]: Is there any way to provide more error information to the user here?

Comment (quartzmo) [reply google-cloud-spanner/lib/google/cloud/spanner/transaction.rb]: Maybe at least a message like: "DML statement is invalid."

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb [hunk 4]
```diff
@@ -314,6 +314,184 @@ def execute sql, params: nil, types: nil, single_use: nil
...
+          results = nil
+          @pool.with_session do |session|
+            results = session.execute \
+              sql, params: params, types: types,
+                   transaction: pdml_transaction(session)
+          end
+          # Stream all PartialResultSet to get ResultSetStats
+          results.rows.to_a
+          # Raise an error if there is not a row count returned
+          raise Google::Cloud::InvalidArgumentError if results.row_count.nil?
```
Comment (quartzmo) [inline google-cloud-spanner/lib/google/cloud/spanner/client.rb]: Add a message here? See `execute_update` comment.

File: google-cloud-spanner/lib/google/cloud/spanner/transaction.rb (hunk 3 above)
Comment (blowmage) [reply google-cloud-spanner/lib/google/cloud/spanner/transaction.rb]: Added.

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb (hunk 4 above)
Comment (blowmage) [reply google-cloud-spanner/lib/google/cloud/spanner/client.rb]: Added.

File: google-cloud-spanner/lib/google/cloud/spanner/transaction.rb [hunk 5]
```diff
@@ -245,13 +246,109 @@ def transaction_id
...
+        #   spanner = Google::Cloud::Spanner.new
+        #   db = spanner.client "my-instance", "my-database"
+        #
+        #   db.transaction do |tx|
+        #     row_count = tx.execute_update(
+        #       "UPDATE users SET name = @name WHERE id = @id",
+        #       params: { id: 1, name: "Charlie" }
+        #     )
+        #   end
+        #
```
Comment (quartzmo) [inline google-cloud-spanner/lib/google/cloud/spanner/transaction.rb]: What do you think about adding an example demonstrating the use of `types`? Here is a unit test for types from this PR:

```rb
row_count = transaction.execute_update "UPDATE users SET project_ids = @list", params: { list: [] }, types: { list: [:INT64] }
```

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb [hunk 6]
```diff
@@ -314,6 +314,187 @@ def execute sql, params: nil, types: nil, single_use: nil
...
+        # @example Query using query parameters:
+        #   require "google/cloud/spanner"
+        #
+        #   spanner = Google::Cloud::Spanner.new
+        #   db = spanner.client "my-instance", "my-database"
+        #
+        #   row_count = db.execute_partition_update \
+        #    "UPDATE users SET friends = NULL WHERE active = @active",
+        #    params: { active: false }
+        #
```
Comment (quartzmo) [inline google-cloud-spanner/lib/google/cloud/spanner/client.rb]: What do you think about adding an example demonstrating the use of `types`? Here is a unit test for types from this PR:

```rb
row_count = client.execute_partition_update "UPDATE users SET project_ids = @list", params: { list: [] }, types: { list: [:INT64] }      
```

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb [hunk 7]
```diff
@@ -314,6 +314,187 @@ def execute sql, params: nil, types: nil, single_use: nil
...
+        # Partitioned DML is an alternate implementation with looser semantics
+        # to enable large-scale changes without running into transaction size
+        # limits or (accidentally) locking the entire table in one large
+        # transaction. At a high level, it partitions the keyspace and executes
+        # the statement on each partition in separate internal transactions.
+        #
+        # Partitioned DML does not guarantee database-wide atomicity of the
+        # statement - it guarantees row-based atomicity, which includes updates
+        # to any indices. Additionally,  it does not guarantee that it will
```
Comment (quartzmo) [inline google-cloud-spanner/lib/google/cloud/spanner/client.rb]: Remove extra space: "Additionally,  it"

File: google-cloud-spanner/lib/google/cloud/spanner/client.rb [hunk 8]
```diff
@@ -314,6 +314,187 @@ def execute sql, params: nil, types: nil, single_use: nil
...
+        # DML in a {Transaction#execute_update} transaction. Smaller scoped
+        # statements, such as an OLTP workload, should prefer using
+        # {Transaction#execute_update}.
+        #
+        # That said, Partitioned DML is not a drop-in replacement for standard
+        # DML used in {Transaction#execute_update}.
+        #
+        # * The DML statement must be fully-partitionable. Specifically, the
+        #   statement must be expressible as the union of many statements which
+        #   each access only a single row of the table.
```
Comment (quartzmo) [inline google-cloud-spanner/lib/google/cloud/spanner/client.rb]: Duplicate. These sentences also appear above. Remove one or the other.
